spec/index.bs

```diff
@@ -4095,6 +4118,123 @@ configuring each of the [=render stages=].
     the fragment shader entry point of the [=pipeline=] and its output colors.
     If it's `null`, the [[#no-color-output]] mode is enabled.
 
+### Primitive clipping ### {#primitive-clipping}
+
+Vertex shaders have to produce a built-in "position", which is a 4-component vector
+of floating-point values denoting the <dfn dfn>clip position</dfn> of a vertex.
+
+Primitives are clipped to the <dfn dfn>clip volume</dfn>, which, for any [=clip position=] |p|
+inside a primitive, is defined by the following set of restrictions:
+  - &minus;|p|.w &le; |p|.x &le; |p|.w
+  - &minus;|p|.w &le; |p|.y &le; |p|.w
+  - 0 &le; |p|.z &le; |p|.w (<dfn dfn>depth clipping</dfn>)
+
+If |descriptor|.{{GPURenderPipelineDescriptor/depthStencil}}.{{GPUDepthStencilState/clampDepth}} is `true`,
+the [=depth clipping=] restriction of the [=clip volume=] is not applied.
+
+A primitive passes through this stage unchanged if every one of its edges
+lie entirely inside the [=clip volume=].
+If the edges of a primitives intersect the boundary of the [=clip volume=],
+the intersecting edges are reconnected by new edges that lie along the boundary of the [=clip volume=].
+For triangular primitives (|descriptor|.{{GPURenderPipelineDescriptor/primitive}}.{{GPUPrimitiveState/topology}} is
+{{GPUPrimitiveTopology/"triangle-list"}} or {{GPUPrimitiveTopology/"triangle-strip"}}), this reconnection
+may result in introduction of new vertices into the polygon, internally.
```

Comment:
Out of curiosity, is that true in newer GPU hardware that has VS / FS on the GPU? It seems it would be easier to just do the interpolation in the rasterizer and discard the fragments. Or do we expect actual WebGPU hardware to run extra invocations of the VS?

Reply:
You can't talk about rasterizer until you clip the VS-output positions, since there is no screen space before it.
No extra invocations are done. This stage is after the VS stage. It's basically a bunch of intermediate vertices.

Reply:
Thinking about it, I'm fairly convinced that's not how hardware works anymore. Instead FrontFace / cullmode is done for the whole triangle, potentially before clipping. Then clipping + clip planes happen with some kind of coarse scan line algorithm. Then for each smaller tile, finer rasterization is done by testing each fragment against the various planes defining the triangle.

Finally the fragments inputs are computed by interpolating from the triangle's vertices. It does conceptually the exact same thing as what's described here but doesn't need the complexity of generating new triangles to form a polygon.

Reply:
The algorithm I described is basically what Vulkan spec has, just in different words, and with some issues fixed.
What you are describing seems a bit far-fetched. From the discussions I recently had, the HW still does primitive clipping, at the very minimum - for the near plane. So the spec text is close to what the HW is doing.

Reply:
It just feel unnecessarily complicated given it's all invisible to the application and not clear if HW follows it (the Vulkan spec probably copy-pasted stuff from the GL spec that copy pasted it version from version from GL 2.1). How about we land this direction and I later make a PR with a proposal for restructuration that we can discuss?

Reply:
Yes, please feel free to follow-up, of course!
I'm not sure if this simplification will stand, but we don't need to block this PR.